crates/lib3h/src/rrdht_util/r_value.rs

```diff
@@ -0,0 +1,341 @@
+}
+
+impl Default for RValuePeerRecord {
+    fn default() -> Self {
+        Self {
+            storage_arc: Arc::new(0.into(), 0),
+            uptime_0_to_1: 0.0,
+        }
+    }
+}
+
+impl RValuePeerRecord {
+    /// what arc is this agent claiming to store
+    pub fn storage_arc(mut self, storage_arc: Arc) -> Self {
+        self.storage_arc = storage_arc;
+        self
+    }
+
+    /// what is this agent's uptime
+    /// - 0 = never reachable
+    /// - 1 = always reachable
+    pub fn uptime_0_to_1(mut self, uptime_0_to_1: f64) -> Self {
+        self.uptime_0_to_1 = uptime_0_to_1;
+        self
+    }
+}
+
+/// builder for a set of peers we know about
+/// this should generally either be our entire storage arc of peer info
+/// or, if it is huge, small sample arc
+pub struct RValuePeerRecordSet {
+    arc_of_included_peer_records: Arc,
+    peer_records: Vec<RValuePeerRecord>,
+}
+
+impl Default for RValuePeerRecordSet {
+    fn default() -> Self {
+        Self {
+            arc_of_included_peer_records: Arc::new(0.into(), 0),
+            peer_records: Vec::new(),
+        }
+    }
+}
+
+impl RValuePeerRecordSet {
+    /// the arc we are representing with this PeerRecordSet
+    pub fn arc_of_included_peer_records(mut self, arc_of_included_peer_records: Arc) -> Self {
+        self.arc_of_included_peer_records = arc_of_included_peer_records;
+        self
+    }
+
+    /// push a new PeerRecord into this record set
+    pub fn push_peer_record(mut self, peer_record: RValuePeerRecord) -> Self {
+        if !self
+            .arc_of_included_peer_records
+            .contains_location(peer_record.storage_arc.center())
+        {
+            panic!("peer record location is not located within arc_of_included_peer_records");
+        }
+        self.peer_records.push(peer_record);
+        self
+    }
+
+    /// guess the total node count, based on our known arc / peer count
+    pub fn guess_total_node_count(&self) -> f64 {
+        self.peer_records.len() as f64
+            * (ARC_LENGTH_MAX as f64 / self.arc_of_included_peer_records.length() as f64)
+    }
+}
+
+/// Interpolate qualified redundancy in the network given our current
+/// knowledge of peers' claimed storage arcs.
+/// For a given storage arc (should match *this* agent's storage arc
+/// and therefore we should know about all peers here (minus some eventuals)),
+/// if all the peers we know about are actually storing what they claim to be
+/// we can interpolate what the whole network "R" value is
+/// also taking into account our experience of those agents' reachability.
+pub fn interpolate_r_value_for_given_arc(peer_record_set: &RValuePeerRecordSet) -> f64 {
+    let mut running_avg_coverage = 0.0_f64;
+
+    for (i, record) in peer_record_set.peer_records.iter().enumerate() {
+        // what is the total coverage factor for this record?
+        let coverage =
+            (record.storage_arc.length() as f64 / ARC_LENGTH_MAX as f64) * record.uptime_0_to_1;
```

Comment:
Weighted linearly by uptime. Interesting heuristic but I think this is a great candidate for a ML system that considers uptime percentage as just one of several features (network latency could be another). 

For now I think this is perfectly fine but it definitely needs to be tested, especially around the nearly zero uptime boundary. Is there a paper I haven't read that establishes convergence properties of this network regarding coverage? Intuitively to me how much you store is not so strongly dependent on uptime, but in the above equation it certainly is.